From 0b45884eb0a6576d588e9181d4a7282bc731b1bd Mon Sep 17 00:00:00 2001
From: "Nguyen Van Linh [FGA.BU13]" <LinhNv25@fsoft.com.vn>
Date: Wed, 14 Nov 2018 16:20:25 +0700
Subject: [PATCH 248/628] drm: rcar-du: Fix mode_fixup callback function

It is prohibition to set up the width of the multiple of 16 for
the specification of H/W in R-Car series. Therefore, when it was
the width which is not a multiple of 16, the procedure which
returns an error was added.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Hung Dong <hung.dong.xd@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNv25@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index f9d5314..a527874 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -558,7 +558,16 @@ static bool rcar_du_crtc_mode_fixup(struct drm_crtc *crtc,
 				    const struct drm_display_mode *mode,
 				    struct drm_display_mode *adjusted_mode)
 {
-	/* TODO Fixup modes */
+	struct rcar_du_crtc *rcrtc = to_rcar_crtc(crtc);
+	struct rcar_du_device *rcdu = rcrtc->group->dev;
+
+	/* It is prohibition to set up the width of the multiple of 16 */
+	/* for the specification of H/W in R-Car series */
+	if ((mode) && (mode->hdisplay % 16)) {
+		dev_err(rcdu->dev,
+			"Error! width of the multiple of 16 is prohibition\n");
+		return false;
+	}
 	return true;
 }
 
-- 
2.7.4

